src/dictionary.c

```diff
@@ -22,66 +23,105 @@ int32_t ceil2(int32_t x) {
 
 // Dictonary object ------------------------------------------------------------
 
-static struct dictionary* new_dictionary_impl(SEXP x, bool partial);
+static struct dictionary* new_dictionary_opts(SEXP x, struct dictionary_opts* opts);
 
 
-static int nil_equal(const void* x, R_len_t i, const void* y, R_len_t j) {
+static int nil_p_equal(const void* x, R_len_t i, const void* y, R_len_t j) {
```

Comment:
Should all of these functions be `inline`d? Not sure if it would make a difference but it seems like the right thing to do here

Reply:
I don't think it makes sense here since there is no inlining possible. These functions are accessed through pointers stored in the dictionary structure.

Reply:
Ah that makes sense